assets/manpages/hanko.1
.ie \n(.g .ds Aq \(aq
.el .ds Aq '
.TH hanko 1  "hanko " 
.SH NAME
hanko \- Keeps your Git allowed signers file up to date with signing keys configured on software development platforms like GitHub and GitLab.
.SH SYNOPSIS
\fBhanko\fR [\fB\-v\fR|\fB\-\-verbose\fR]... [\fB\-c\fR|\fB\-\-config\fR] [\fB\-\-file\fR] [\fB\-h\fR|\fB\-\-help\fR] <\fIsubcommands\fR>
.SH DESCRIPTION
Keeps your Git allowed signers file up to date with signing keys configured on software development platforms like GitHub and GitLab.
.SH OPTIONS
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Use verbose output
.TP
\fB\-c\fR, \fB\-\-config\fR=\fIPATH\fR [default: ~/.config/hanko/config.toml]
The configuration file
.RS
May also be specified with the \fBHANKO_CONFIG\fR environment variable. 
.RE
.TP
\fB\-\-file\fR=\fIPATH\fR
The allowed signers file
.RS
May also be specified with the \fBHANKO_ALLOWED_SIGNERS\fR environment variable. 
.RE
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help
.SH SUBCOMMANDS
.TP
hanko\-update(1)
Update the allowed signers file
.TP
hanko\-signer(1)
Manage allowed signers


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.5.2] - 2024-12-15

### Added

- Shell completion files for bash, zsh, fish, elvish and powershell.
- Manpages.

### Changed

- Version output now contains features and build time information.

### Internal

- Minimize enabled features for external dependencies and improve binary size.

## [0.5.1] - 2024-11-24

### Added

- Debian Packages to releases.

### Internal

- Updated all dependencies.

## [0.5.0] - 2024-11-23

### Added

- Subcommand to add allowed signers.

### Changed

- Unknown fields in the configuration will now cause an error instead of being ignored.
- Loading a configuration file that does not exist will cause an error.
- All of a signers sources are now queried concurrently.
- If a user does not exist on a source a warning will be logged instead of returning an error.
- If a user does not have any keys configured on a source, a warning will be logged.

## [0.4.1] - 2024-10-29

### Security

- Update yanked [futures-rs](https://github.com/rust-lang/futures-rs) version containing use after free.

## [0.4.0] - 2024-10-28

### Added

- This Changelog.
- Build provenance file to releases.

### Removed

- Allowed signers file option from configuration file.

## [0.3.0] - 2024-10-25

### Added

- Attested container image built in CI.
- Improved logging.
- Using GitHub as default source when left unspecified.
- Tracking of test coverage.
- Logging GitHub ratelimit details when using the trace log level.

### Changed

- Make gix-config an optional dependency enabled per default by the `detect-allowed-signers` feature.

### Internal

- Improve crate structure and naming.
- Use reference counting for sources to enable usage from different threads.

## [0.2.1] - 2024-07-18

### Security

- Update gix-path to fix CVE-2024-40644.

## [0.2.0] - 2024-07-18

### Added

- Logging of API responses in trace mode.
- MIT license.

### Fixed

- aarch64 builds.

[unreleased]: https://github.com/SRv6d/hanko/compare/v0.5.2...HEAD
[0.5.2]: https://github.com/SRv6d/hanko/compare/v0.5.1...v0.5.2
[0.5.1]: https://github.com/SRv6d/hanko/compare/v0.5.0...v0.5.1
[0.5.0]: https://github.com/SRv6d/hanko/compare/v0.4.1...v0.5.0
[0.4.1]: https://github.com/SRv6d/hanko/compare/v0.4.0...v0.4.1
[0.4.0]: https://github.com/SRv6d/hanko/compare/v0.3.0...v0.4.0
[0.3.0]: https://github.com/SRv6d/hanko/compare/v0.2.1...v0.3.0
[0.2.1]: https://github.com/SRv6d/hanko/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com/SRv6d/hanko/compare/v0.1.0...v0.2.0


assets/completions/hanko.bash
_hanko() {
    local i cur prev opts cmd
    COMPREPLY=()
    cur="${COMP_WORDS[COMP_CWORD]}"
    prev="${COMP_WORDS[COMP_CWORD-1]}"
    cmd=""
    opts=""

    for i in ${COMP_WORDS[@]}
    do
        case "${cmd},${i}" in
            ",$1")
                cmd="hanko"
                ;;
            hanko,help)
                cmd="hanko__help"
                ;;
            hanko,signer)
                cmd="hanko__signer"
                ;;
            hanko,update)
                cmd="hanko__update"
                ;;
            hanko__help,help)
                cmd="hanko__help__help"
                ;;
            hanko__help,signer)
                cmd="hanko__help__signer"
                ;;
            hanko__help,update)
                cmd="hanko__help__update"
                ;;
            hanko__help__signer,add)
                cmd="hanko__help__signer__add"
                ;;
            hanko__signer,add)
                cmd="hanko__signer__add"
                ;;
            hanko__signer,help)
                cmd="hanko__signer__help"
                ;;
            hanko__signer__help,add)
                cmd="hanko__signer__help__add"
                ;;
            hanko__signer__help,help)
                cmd="hanko__signer__help__help"
                ;;
            *)
                ;;
        esac
    done

    case "${cmd}" in
        hanko)
            opts="-c -v -h -V --config --file --verbose --help --version update signer help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 1 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                --config)
                    local oldifs
                    if [ -n "${IFS+x}" ]; then
                        oldifs="$IFS"
                    fi
                    IFS=$'\n'
                    COMPREPLY=($(compgen -f "${cur}"))
                    if [ -n "${oldifs+x}" ]; then
                        IFS="$oldifs"
                    fi
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o filenames
                    fi
                    return 0
                    ;;
                -c)
                    local oldifs
                    if [ -n "${IFS+x}" ]; then
                        oldifs="$IFS"
                    fi
                    IFS=$'\n'
                    COMPREPLY=($(compgen -f "${cur}"))
                    if [ -n "${oldifs+x}" ]; then
                        IFS="$oldifs"
                    fi
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o filenames
                    fi
                    return 0
                    ;;
                --file)
                    local oldifs
                    if [ -n "${IFS+x}" ]; then
                        oldifs="$IFS"
                    fi
                    IFS=$'\n'
                    COMPREPLY=($(compgen -f "${cur}"))
                    if [ -n "${oldifs+x}" ]; then
                        IFS="$oldifs"
                    fi
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o filenames
                    fi
                    return 0
                    ;;
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        hanko__help)
            opts="update signer help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        hanko__help__help)
            opts=""
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        hanko__help__signer)
            opts="add"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        hanko__help__signer__add)
            opts=""
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 4 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        hanko__help__update)
            opts=""
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        hanko__signer)
            opts="-c -v -h --config --file --verbose --help add help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                --config)
                    local oldifs
                    if [ -n "${IFS+x}" ]; then
                        oldifs="$IFS"
                    fi
                    IFS=$'\n'
                    COMPREPLY=($(compgen -f "${cur}"))
                    if [ -n "${oldifs+x}" ]; then
                        IFS="$oldifs"
                    fi
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o filenames
                    fi
                    return 0
                    ;;
                -c)
                    local oldifs
                    if [ -n "${IFS+x}" ]; then
                        oldifs="$IFS"
                    fi
                    IFS=$'\n'
                    COMPREPLY=($(compgen -f "${cur}"))
                    if [ -n "${oldifs+x}" ]; then
                        IFS="$oldifs"
                    fi
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o filenames
                    fi
                    return 0
                    ;;
                --file)
                    local oldifs
                    if [ -n "${IFS+x}" ]; then
                        oldifs="$IFS"
                    fi
                    IFS=$'\n'
                    COMPREPLY=($(compgen -f "${cur}"))
                    if [ -n "${oldifs+x}" ]; then
                        IFS="$oldifs"
                    fi
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o filenames
                    fi
                    return 0
                    ;;
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        hanko__signer__add)
            opts="-s -c -v -h --source --no-update --config --file --verbose --help <NAME> [PRINCIPALS]..."
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                --source)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                -s)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --config)
                    local oldifs
                    if [ -n "${IFS+x}" ]; then
                        oldifs="$IFS"
                    fi
                    IFS=$'\n'
                    COMPREPLY=($(compgen -f "${cur}"))
                    if [ -n "${oldifs+x}" ]; then
                        IFS="$oldifs"
                    fi
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o filenames
                    fi
                    return 0
                    ;;
                -c)
                    local oldifs
                    if [ -n "${IFS+x}" ]; then
                        oldifs="$IFS"
                    fi
                    IFS=$'\n'
                    COMPREPLY=($(compgen -f "${cur}"))
                    if [ -n "${oldifs+x}" ]; then
                        IFS="$oldifs"
                    fi
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o filenames
                    fi
                    return 0
                    ;;
                --file)
                    local oldifs
                    if [ -n "${IFS+x}" ]; then
                        oldifs="$IFS"
                    fi
                    IFS=$'\n'
                    COMPREPLY=($(compgen -f "${cur}"))
                    if [ -n "${oldifs+x}" ]; then
                        IFS="$oldifs"
                    fi
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o filenames
                    fi
                    return 0
                    ;;
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        hanko__signer__help)
            opts="add help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        hanko__signer__help__add)
            opts=""
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 4 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        hanko__signer__help__help)
            opts=""
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 4 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        hanko__update)
            opts="-c -v -h --config --file --verbose --help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                --config)
                    local oldifs
                    if [ -n "${IFS+x}" ]; then
                        oldifs="$IFS"
                    fi
                    IFS=$'\n'
                    COMPREPLY=($(compgen -f "${cur}"))
                    if [ -n "${oldifs+x}" ]; then
                        IFS="$oldifs"
                    fi
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o filenames
                    fi
                    return 0
                    ;;
                -c)
                    local oldifs
                    if [ -n "${IFS+x}" ]; then
                        oldifs="$IFS"
                    fi
                    IFS=$'\n'
                    COMPREPLY=($(compgen -f "${cur}"))
                    if [ -n "${oldifs+x}" ]; then
                        IFS="$oldifs"
                    fi
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o filenames
                    fi
                    return 0
                    ;;
                --file)
                    local oldifs
                    if [ -n "${IFS+x}" ]; then
                        oldifs="$IFS"
                    fi
                    IFS=$'\n'
                    COMPREPLY=($(compgen -f "${cur}"))
                    if [ -n "${oldifs+x}" ]; then
                        IFS="$oldifs"
                    fi
                    if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
                        compopt -o filenames
                    fi
                    return 0
                    ;;
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
    esac
}

if [[ "${BASH_VERSINFO[0]}" -eq 4 && "${BASH_VERSINFO[1]}" -ge 4 || "${BASH_VERSINFO[0]}" -gt 4 ]]; then
    complete -F _hanko -o nosort -o bashdefault -o default hanko
else
    complete -F _hanko -o bashdefault -o default hanko
fi
